fn/commands/deploy.go

```diff
@@ -0,0 +1,241 @@
+package commands
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/iron-io/functions/fn/common"
+	functions "github.com/iron-io/functions_go"
+	"github.com/urfave/cli"
+)
+
+func Deploy() cli.Command {
+	cmd := deploycmd{
+		RoutesApi: functions.NewRoutesApi(),
+	}
+	var flags []cli.Flag
+	flags = append(flags, cmd.flags()...)
+	return cli.Command{
+		Name:      "deploy",
+		ArgsUsage: "<appName>",
+		Usage:     "scan local directory for functions, build and push all of them to `APPNAME`.",
+		Flags:     flags,
+		Action:    cmd.scan,
+	}
+}
+
+type deploycmd struct {
+	appName string
+	*functions.RoutesApi
+
+	wd          string
+	verbose     bool
+	incremental bool
+	skippush    bool
+
+	verbwriter io.Writer
+}
+
+func (p *deploycmd) flags() []cli.Flag {
+	return []cli.Flag{
+		cli.BoolFlag{
+			Name:        "v",
+			Usage:       "verbose mode",
+			Destination: &p.verbose,
+		},
+		cli.StringFlag{
+			Name:        "d",
+			Usage:       "working directory",
+			Destination: &p.wd,
+			EnvVar:      "WORK_DIR",
+			Value:       "./",
+		},
+		cli.BoolFlag{
+			Name:        "i",
+			Usage:       "uses incremental building",
+			Destination: &p.incremental,
+		},
+		cli.BoolFlag{
+			Name:        "skip-push",
+			Usage:       "does not push Docker built images onto Docker Hub - useful for local development.",
+			Destination: &p.skippush,
+		},
+	}
+}
+
+func (p *deploycmd) scan(c *cli.Context) error {
+	p.appName = c.Args().First()
+	p.verbwriter = common.Verbwriter(p.verbose)
+
+	var walked bool
+
+	err := filepath.Walk(p.wd, func(path string, info os.FileInfo, err error) error {
+		if path != p.wd && info.IsDir() {
+			return filepath.SkipDir
+		}
+
+		if !isFuncfile(path, info) {
+			return nil
+		}
+
+		if p.incremental && !isstale(path) {
+			return nil
+		}
+
+		e := p.deploy(path)
+		if err != nil {
+			fmt.Fprintln(p.verbwriter, path, e)
+		}
+
+		now := time.Now()
+		os.Chtimes(path, now, now)
+		walked = true
+		return e
+	})
+	if err != nil {
+		fmt.Fprintf(p.verbwriter, "file walk error: %s\n", err)
+	}
+
+	if !walked {
+		return errors.New("No function file found.")
+	}
+
+	return nil
+}
+
+// deploy will take the found function and check for the presence of a
+// Dockerfile, and run a three step process: parse functions file, build and
+// push the container, and finally it will update function's route. Optionally,
+// the route can be overriden inside the functions file.
+func (p *deploycmd) deploy(path string) error {
+	fmt.Fprintln(p.verbwriter, "deploying", path)
+
+	funcfile, err := common.Buildfunc(p.verbwriter, path)
+	if err != nil {
+		return err
+	}
+
+	if p.skippush {
```

Comment:
So if `skippush` is enabled then routes will not updated ?

Reply:
Thanks for the review ! This part is as is with master, I didn't change any logic within functions. Yes, it seems original intention is that if you pass skippush, routes will not get updated. I am not sure if this is intentional, or a bug.

https://github.com/iron-io/functions/blob/master/fn/deploy.go#L123